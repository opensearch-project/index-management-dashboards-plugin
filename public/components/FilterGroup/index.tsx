import { EuiFilterButton, EuiFilterButtonProps, EuiFilterGroup, EuiFilterSelectItem, EuiPopover } from "@elastic/eui";
import React, { useState } from "react";

export interface IFilterGroupProps {
  options: { label: string }[];
  value?: string[];
  filterButtonProps?: EuiFilterButtonProps;
  useNewUx?: boolean;
  onChange?: (val: IFilterGroupProps["value"]) => void;
}

export default function FilterGroup({ options, value, filterButtonProps, useNewUx, onChange }: IFilterGroupProps) {
  const [isPopoverOpen, setIsPopoverOpen] = useState(false);

  const onButtonClick = () => {
    setIsPopoverOpen(!isPopoverOpen);
  };

  const closePopover = () => {
    setIsPopoverOpen(false);
  };

  return (
    <EuiFilterGroup>
      <EuiPopover
        button={
          <EuiFilterButton
            iconType="arrowDown"
            onClick={onButtonClick}
            isSelected={isPopoverOpen}
            numFilters={options?.length}
            hasActiveFilters={!!value?.length}
            numActiveFilters={value?.length}
            size={useNewUx ? "s" : undefined}
            {...filterButtonProps}
          />
        }
        isOpen={isPopoverOpen}
        closePopover={closePopover}
        panelPaddingSize="none"
      >
        <div className="ouiFilterSelect__items">
          {options.map((item, index) => (
            <EuiFilterSelectItem
              checked={value?.includes(item.label) ? "on" : "off"}
              key={item.label}
              data-test-subj={`FilterGroupSelectItem-${item.label}`}
              onClick={() => {
                const findIndex = (value || []).findIndex((valueItem) => valueItem === item.label);
                const finalValue = value || [];
                if (findIndex > -1) {
                  finalValue.splice(findIndex, 1);
                } else {
                  finalValue.push(item.label);
                }
                onChange?.([...finalValue]);
              }}
            >
              {item.label}
            </EuiFilterSelectItem>
          ))}
        </div>
      </EuiPopover>
    </EuiFilterGroup>
  );
}
