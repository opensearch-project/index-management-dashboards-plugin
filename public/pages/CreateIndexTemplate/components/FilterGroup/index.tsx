import { EuiFilterButton, EuiFilterButtonProps, EuiFilterGroup, EuiFilterSelectItem, EuiLoadingChart, EuiPopover } from "@elastic/eui";
import React, { useState } from "react";
interface IFilterGroupProps {
  options: { label: string }[];
  value?: string[];
  filterButtonProps?: EuiFilterButtonProps;
  onChange?: (val: IFilterGroupProps["value"]) => void;
}

export default function FilterGroup({ options, value, filterButtonProps, onChange }: IFilterGroupProps) {
  const [isPopoverOpen, setIsPopoverOpen] = useState(false);

  const onButtonClick = () => {
    setIsPopoverOpen(!isPopoverOpen);
  };

  const closePopover = () => {
    setIsPopoverOpen(false);
  };

  return (
    <EuiFilterGroup>
      <EuiPopover
        button={
          <EuiFilterButton
            iconType="arrowDown"
            onClick={onButtonClick}
            isSelected={isPopoverOpen}
            numFilters={options?.length}
            hasActiveFilters={!!value?.length}
            numActiveFilters={value?.length}
            {...filterButtonProps}
          />
        }
        isOpen={isPopoverOpen}
        closePopover={closePopover}
        panelPaddingSize="none"
      >
        <div className="ouiFilterSelect__items">
          {options.map((item, index) => (
            <EuiFilterSelectItem
              checked={value?.includes(item.label) ? "on" : "off"}
              key={item.label}
              onClick={() => {
                const findIndex = (value || []).findIndex((valueItem) => valueItem === item.label);
                if (findIndex > -1) {
                  value?.splice(findIndex, 1);
                } else {
                  value?.push(item.label);
                }
                onChange?.([...(value || [])]);
              }}
            >
              {item.label}
            </EuiFilterSelectItem>
          ))}
        </div>
      </EuiPopover>
    </EuiFilterGroup>
  );
}
