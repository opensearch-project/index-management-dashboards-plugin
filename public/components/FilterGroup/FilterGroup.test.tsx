/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useState } from "react";
import { act, render, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import FilterGroup, { IFilterGroupProps } from "./index";

const WrappedComponent = (props: IFilterGroupProps) => {
  const [value, onChange] = useState<string[] | undefined>();
  return <FilterGroup {...props} value={value} onChange={(val) => onChange(val || [])} />;
};

describe("<FilterGroup /> spec", () => {
  it("render the component", async () => {
    const { findByPlaceholderText, getByTestId, findByText, getByText, queryByText } = render(
      <>
        <WrappedComponent
          options={[{ label: "test option" }]}
          value={[]}
          filterButtonProps={{
            placeholder: "test",
            "data-test-subj": "test",
          }}
        />
        <button>hide</button>
      </>
    );
    await findByPlaceholderText("test");
    expect(document.body.children).toMatchSnapshot();
    await userEvent.click(getByTestId("test"));
    await findByText("test option");
    expect(document.body.children).toMatchSnapshot();
    await userEvent.click(getByText("test option"));
    await waitFor(() => {
      expect(document.querySelector("[aria-label='1 active filters']")).toBeInTheDocument();
      expect(document.body.children).toMatchSnapshot();
    });
    await userEvent.click(getByText("test option"));
    await waitFor(() => {
      expect(document.querySelector("[aria-label='1 available filters']")).toBeInTheDocument();
      expect(document.body.children).toMatchSnapshot();
    });
    act(() => {
      userEvent.click(getByText("hide"));
    });
    await waitFor(() => {
      expect(queryByText("test option")).toBeNull();
    });
  });
});
