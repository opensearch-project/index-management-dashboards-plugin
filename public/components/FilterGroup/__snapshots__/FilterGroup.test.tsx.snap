// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilterGroup /> spec render the component 1`] = `
HTMLCollection [
  <div>
    <div
      class="euiFilterGroup"
    >
      <div
        class="euiPopover euiPopover--anchorDownCenter"
      >
        <div
          class="euiPopover__anchor"
        >
          <button
            class="euiButtonEmpty euiButtonEmpty--text euiFilterButton euiFilterButton-hasNotification euiFilterButton--hasIcon"
            data-test-subj="test"
            placeholder="test"
            type="button"
          >
            <span
              class="euiButtonContent euiButtonContent--iconRight euiButtonEmpty__content"
            >
              EuiIconMock
              <span
                class="euiButtonEmpty__text euiFilterButton__text-hasNotification"
              >
                <span
                  class="euiFilterButton__textShift"
                />
                <span
                  aria-label="1 available filters"
                  class="euiNotificationBadge euiNotificationBadge--medium euiNotificationBadge--subdued euiFilterButton__notification"
                >
                  1
                </span>
              </span>
            </span>
          </button>
        </div>
      </div>
    </div>
    <button>
      hide
    </button>
  </div>,
]
`;

exports[`<FilterGroup /> spec render the component 2`] = `
HTMLCollection [
  <div
    aria-hidden="true"
    data-aria-hidden="true"
  >
    <div
      class="euiFilterGroup"
    >
      <div
        class="euiPopover euiPopover--anchorDownCenter euiPopover-isOpen"
      >
        <div
          class="euiPopover__anchor"
        >
          <button
            class="euiButtonEmpty euiButtonEmpty--text euiFilterButton euiFilterButton-isSelected euiFilterButton-hasNotification euiFilterButton--hasIcon"
            data-test-subj="test"
            placeholder="test"
            type="button"
          >
            <span
              class="euiButtonContent euiButtonContent--iconRight euiButtonEmpty__content"
            >
              EuiIconMock
              <span
                class="euiButtonEmpty__text euiFilterButton__text-hasNotification"
              >
                <span
                  class="euiFilterButton__textShift"
                />
                <span
                  aria-label="1 available filters"
                  class="euiNotificationBadge euiNotificationBadge--medium euiNotificationBadge--subdued euiFilterButton__notification"
                >
                  1
                </span>
              </span>
            </span>
          </button>
        </div>
      </div>
    </div>
    <button>
      hide
    </button>
  </div>,
  <div>
    <div
      aria-hidden="true"
      data-aria-hidden="true"
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="0"
    />
    <div
      data-focus-lock-disabled="false"
    >
      <div
        aria-describedby="some_html_id"
        aria-live="off"
        aria-modal="true"
        class="euiPanel euiPanel--borderRadiusMedium euiPanel--plain euiPanel--noShadow euiPopover__panel euiPopover__panel--bottom euiPopover__panel-isOpen"
        data-autofocus="true"
        role="dialog"
        style="top: 16px; left: -22px; z-index: 2000;"
        tabindex="0"
      >
        <div
          class="euiPopover__panelArrow euiPopover__panelArrow--bottom"
          style="left: 10px; top: 0px;"
        />
        <p
          class="euiScreenReaderOnly"
          id="some_html_id"
        >
          You are in a dialog. To close this dialog, hit escape.
        </p>
        <div>
          <div
            class="ouiFilterSelect__items"
          >
            <button
              class="euiFilterSelectItem"
              data-test-subj="FilterGroupSelectItem-test option"
              role="option"
              type="button"
            >
              <span
                class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow"
              >
                <div
                  class="euiFlexItem euiFlexItem--flexGrowZero"
                >
                  EuiIconMock
                </div>
                <span
                  class="euiFlexItem euiFilterSelectItem__content"
                >
                  test option
                </span>
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      aria-hidden="true"
      data-aria-hidden="true"
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="0"
    />
  </div>,
]
`;

exports[`<FilterGroup /> spec render the component 3`] = `
HTMLCollection [
  <div
    aria-hidden="true"
    data-aria-hidden="true"
  >
    <div
      class="euiFilterGroup"
    >
      <div
        class="euiPopover euiPopover--anchorDownCenter euiPopover-isOpen"
      >
        <div
          class="euiPopover__anchor"
        >
          <button
            class="euiButtonEmpty euiButtonEmpty--text euiFilterButton euiFilterButton-isSelected euiFilterButton-hasActiveFilters euiFilterButton-hasNotification euiFilterButton--hasIcon"
            data-test-subj="test"
            placeholder="test"
            type="button"
          >
            <span
              class="euiButtonContent euiButtonContent--iconRight euiButtonEmpty__content"
            >
              EuiIconMock
              <span
                class="euiButtonEmpty__text euiFilterButton__text-hasNotification"
              >
                <span
                  class="euiFilterButton__textShift"
                />
                <span
                  aria-label="1 active filters"
                  class="euiNotificationBadge euiNotificationBadge--medium euiFilterButton__notification"
                >
                  1
                </span>
              </span>
            </span>
          </button>
        </div>
      </div>
    </div>
    <button>
      hide
    </button>
  </div>,
  <div>
    <div
      aria-hidden="true"
      data-aria-hidden="true"
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="0"
    />
    <div
      data-focus-lock-disabled="false"
    >
      <div
        aria-describedby="some_html_id"
        aria-live="off"
        aria-modal="true"
        class="euiPanel euiPanel--borderRadiusMedium euiPanel--plain euiPanel--noShadow euiPopover__panel euiPopover__panel--bottom euiPopover__panel-isOpen"
        data-autofocus="true"
        role="dialog"
        style="top: 16px; left: -22px; z-index: 2000;"
        tabindex="0"
      >
        <div
          class="euiPopover__panelArrow euiPopover__panelArrow--bottom"
          style="left: 10px; top: 0px;"
        />
        <p
          class="euiScreenReaderOnly"
          id="some_html_id"
        >
          You are in a dialog. To close this dialog, hit escape.
        </p>
        <div>
          <div
            class="ouiFilterSelect__items"
          >
            <button
              class="euiFilterSelectItem"
              data-test-subj="FilterGroupSelectItem-test option"
              role="option"
              type="button"
            >
              <span
                class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow"
              >
                <div
                  class="euiFlexItem euiFlexItem--flexGrowZero"
                >
                  EuiIconMock
                </div>
                <span
                  class="euiFlexItem euiFilterSelectItem__content"
                >
                  test option
                </span>
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      aria-hidden="true"
      data-aria-hidden="true"
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="0"
    />
  </div>,
]
`;

exports[`<FilterGroup /> spec render the component 4`] = `
HTMLCollection [
  <div
    aria-hidden="true"
    data-aria-hidden="true"
  >
    <div
      class="euiFilterGroup"
    >
      <div
        class="euiPopover euiPopover--anchorDownCenter euiPopover-isOpen"
      >
        <div
          class="euiPopover__anchor"
        >
          <button
            class="euiButtonEmpty euiButtonEmpty--text euiFilterButton euiFilterButton-isSelected euiFilterButton-hasNotification euiFilterButton--hasIcon"
            data-test-subj="test"
            placeholder="test"
            type="button"
          >
            <span
              class="euiButtonContent euiButtonContent--iconRight euiButtonEmpty__content"
            >
              EuiIconMock
              <span
                class="euiButtonEmpty__text euiFilterButton__text-hasNotification"
              >
                <span
                  class="euiFilterButton__textShift"
                />
                <span
                  aria-label="1 available filters"
                  class="euiNotificationBadge euiNotificationBadge--medium euiNotificationBadge--subdued euiFilterButton__notification"
                >
                  1
                </span>
              </span>
            </span>
          </button>
        </div>
      </div>
    </div>
    <button>
      hide
    </button>
  </div>,
  <div>
    <div
      aria-hidden="true"
      data-aria-hidden="true"
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="0"
    />
    <div
      data-focus-lock-disabled="false"
    >
      <div
        aria-describedby="some_html_id"
        aria-live="off"
        aria-modal="true"
        class="euiPanel euiPanel--borderRadiusMedium euiPanel--plain euiPanel--noShadow euiPopover__panel euiPopover__panel--bottom euiPopover__panel-isOpen"
        data-autofocus="true"
        role="dialog"
        style="top: 16px; left: -22px; z-index: 2000;"
        tabindex="0"
      >
        <div
          class="euiPopover__panelArrow euiPopover__panelArrow--bottom"
          style="left: 10px; top: 0px;"
        />
        <p
          class="euiScreenReaderOnly"
          id="some_html_id"
        >
          You are in a dialog. To close this dialog, hit escape.
        </p>
        <div>
          <div
            class="ouiFilterSelect__items"
          >
            <button
              class="euiFilterSelectItem"
              data-test-subj="FilterGroupSelectItem-test option"
              role="option"
              type="button"
            >
              <span
                class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow"
              >
                <div
                  class="euiFlexItem euiFlexItem--flexGrowZero"
                >
                  EuiIconMock
                </div>
                <span
                  class="euiFlexItem euiFilterSelectItem__content"
                >
                  test option
                </span>
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      aria-hidden="true"
      data-aria-hidden="true"
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="0"
    />
  </div>,
]
`;
